### PR TITLE
drivers: spi: esp32xx: add DMA support

### PR DESCRIPTION
**Pull request by LucasTambor**
Add SPI DMA support for esp32/s2/c3.

### CODE REVIEW

File: drivers/spi/spi_esp32_spim.c [hunk 1]
```diff
@@ -55,19 +58,43 @@ static inline void spi_esp32_complete(const struct device *dev,
...
 	spi_hal_dev_config_t *hal_dev = &data->dev_config;
 	spi_hal_trans_config_t *hal_trans = &data->trans_config;
 	size_t chunk_len = spi_context_max_continuous_chunk(&data->ctx);
 	chunk_len = MIN(chunk_len, SOC_SPI_MAXIMUM_BUFFER_SIZE);
+	size_t bit_len = chunk_len << 3;
+	uint8_t *rx_temp = NULL;
+	uint8_t *tx_temp = NULL;
+
+	if (cfg->dma_enabled) {
+		bit_len = (!chunk_len) ? (chunk_len + 1) << 3 : bit_len;
```
Comment (uLipe) [inline drivers/spi/spi_esp32_spim.c:73]: Not so trivial, I would like to see some comment here

File: drivers/spi/spi_esp32_spim.c [hunk 2]
```diff
@@ -55,19 +58,43 @@ static inline void spi_esp32_complete(const struct device *dev,
...
+		if (ctx->tx_buf && !esp_ptr_dma_capable((uint32_t *)&ctx->tx_buf[0])) {
+			tx_temp = k_malloc(((ctx->tx_len << 3) + 7) / 8);
+			if (!tx_temp) {
+				LOG_ERR("Error allocating temp buffer Tx");
+				return -ENOMEM;
+			}
+			memcpy(tx_temp, &ctx->tx_buf[0], ((ctx->tx_len << 3) + 7) / 8);
```
Comment (uLipe) [inline drivers/spi/spi_esp32_spim.c:80]: It would not be better to make this exotic length into a macro?

File: drivers/spi/spi_esp32_spim.c [hunk 3]
```diff
@@ -103,6 +143,53 @@ static void IRAM_ATTR spi_esp32_isr(void *arg)
...
+		return -EIO;
+	}
+
+#ifdef CONFIG_SOC_ESP32C3
+	gdma_hal_init(&data->hal_gdma, 0);
+	gdma_ll_enable_clock(data->hal_gdma.dev, true);
+	gdma_ll_tx_reset_channel(data->hal_gdma.dev, cfg->dma_host);
+	gdma_ll_rx_reset_channel(data->hal_gdma.dev, cfg->dma_host);
+	gdma_ll_tx_connect_to_periph(data->hal_gdma.dev, cfg->dma_host, 0);
+	gdma_ll_rx_connect_to_periph(data->hal_gdma.dev, cfg->dma_host, 0);
```
Comment (uLipe) [inline drivers/spi/spi_esp32_spim.c:163]: Question, is this prepared to the restrictions of GDMA in C3? It has only 3 channels which needs to be arbited and mux'ed with other perioherals.

**Review by uLipe — COMMENTED** [review]
Minor questions, not blockers though

File: drivers/spi/spi_esp32_spim.c (hunk 1 above)
Comment (LucasTambor) [reply drivers/spi/spi_esp32_spim.c]: Simplified the expression and added the comment :)

File: drivers/spi/spi_esp32_spim.c (hunk 2 above)
Comment (LucasTambor) [reply drivers/spi/spi_esp32_spim.c]: Simplified the tx buffer length and added a comment on the rx buffer allocation :)

File: drivers/spi/spi_esp32_spim.c (hunk 3 above)
Comment (LucasTambor) [reply drivers/spi/spi_esp32_spim.c:167]: Not yet, by now this is the only driver that make use of GDMA, so the channel is hard-coded to the host index. When needed we will probably implement a proper driver for GDMA.

**Review by uLipe — DISMISSED** [review]

**Review by sylvioalves — DISMISSED** [review]

**Review by glaubermaroto — DISMISSED** [review]

**Comment by carlescufi** [comment]
@LucasTambor please rebase

**Review by uLipe — APPROVED** [review]

**Review by sylvioalves — APPROVED** [review]

**Review by glaubermaroto — APPROVED** [review]

**Pull request merged**